patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -72,6 +81,84 @@ internal static void Add(ModConfig config)
 		configList2.Add(GeneratePopulatedClone(config));
 	}
 
+	internal static void FinishSetup()
+	{
+		typesWithLocalizationRegistered.Clear();
+		// Register localization for all fields and properties that should show
+		foreach (var activeConfigs in ConfigManager.Configs) {
+			foreach (var config in activeConfigs.Value) {
+				_ = config.DisplayName;
+
+				RegisterLocalizationKeysForMembers(config.GetType());
+			}
+		}
+	}
+
+	private static void RegisterLocalizationKeysForMembers(Type type)
+	{
+		AssemblyManager.GetAssemblyOwner(type.Assembly, out var modName);
+		foreach (PropertyFieldWrapper variable in ConfigManager.GetFieldsAndProperties(type)) {
+#pragma warning disable CS0618
+			if (Attribute.IsDefined(variable.MemberInfo, typeof(JsonIgnoreAttribute)) && !(Attribute.IsDefined(variable.MemberInfo, typeof(LabelAttribute)) || Attribute.IsDefined(variable.MemberInfo, typeof(ShowDespiteJsonIgnoreAttribute))))
+				continue;
+#pragma warning restore CS0618
+
+			try {
+				RegisterLocalizationKeysForMemberType(variable.Type, type.Assembly);
+
+				// Handle obsolete attributes. Use them to populate value of key, if present, to ease porting.
+#pragma warning disable CS0618 // Type or member is obsolete
+				var labelObsolete = (LabelAttribute)Attribute.GetCustomAttribute(variable.MemberInfo, typeof(LabelAttribute));
+				var tooltipObsolete = (TooltipAttribute)Attribute.GetCustomAttribute(variable.MemberInfo, typeof(TooltipAttribute));
+#pragma warning restore CS0618 // Type or member is obsolete
+
+				// Label and Tooltip will always exist. Header is optional, need to be used to exist.
+				var header = GetLocalizedHeader(variable, throwErrors: true);
+				if (header != null) {
+					string identifier = header.IsIdentifier ? header.identifier : variable.Name;
+					Language.GetOrRegister(header.key, () => $"{Regex.Replace(identifier, "([A-Z])", " $1").Trim()} Header");
+				}
+
+				string labelKey = GetConfigKey<LabelKeyAttribute>(variable.MemberInfo, dataName: "Label");
+				Language.GetOrRegister(labelKey, () => labelObsolete?.LocalizationEntry ?? Regex.Replace(variable.Name, "([A-Z])", " $1").Trim());
+
+				if (!type.IsEnum) {
+					string tooltipKey = GetConfigKey<TooltipKeyAttribute>(variable.MemberInfo, dataName: "Tooltip");
+					Language.GetOrRegister(tooltipKey, () => tooltipObsolete?.LocalizationEntry ?? "");
+				}
+			}
+			catch (ValueNotTranslationKeyException e) when (!e.handled) {
+				if (e.errorOnType) {
+					e.additional = $"The class '{variable.Type.FullName}' caused this exception.";
+				}
+				else {
+					e.additional = $"The member '{variable.Name}' found in the '{variable.MemberInfo.DeclaringType}' class caused this exception.";
+				}
+				e.Data["mod"] = modName;
+				e.handled = true; // Recursion will cause this to be called higher in the stack, causing issues.
```

Comment:
I have the impression that we can get this info, or just as good info at the place we throw the exception. Which would remove the need for the recursion bailout. 

`e.Data["mod"] = modName;` is pretty standard, but that could go in the `FinishSetup` loop (per mod)